spark3-extensions/src/main/scala/org/apache/spark/sql/catalyst/analysis/AlignMergeIntoTable.scala

```diff
@@ -0,0 +1,91 @@
+
+import org.apache.spark.sql.AnalysisException
+import org.apache.spark.sql.catalyst.plans.logical.{Assignment, DeleteAction, InsertAction, LogicalPlan, MergeIntoTable, UpdateAction}
+import org.apache.spark.sql.catalyst.rules.Rule
+import org.apache.spark.sql.internal.SQLConf
+
+case class AlignMergeIntoTable(conf: SQLConf) extends Rule[LogicalPlan] with AssignmentAlignmentSupport {
+
+  override def apply(plan: LogicalPlan): LogicalPlan = plan resolveOperators {
+    case m: MergeIntoTable if m.resolved =>
+      val alignedMatchedActions = m.matchedActions.map {
+        case u @ UpdateAction(_, assignments) =>
+          u.copy(assignments = alignAssignments(m.targetTable, assignments))
+        case d: DeleteAction => d
+      }
+
+      val alignedNotMatchedActions = m.notMatchedActions.map {
+        case i @ InsertAction(_, assignments) =>
+          // check no nested columns are present
+          val namePartsSeq = assignments.map(_.key).map(getNameParts)
+          namePartsSeq.foreach { nameParts =>
+            if (nameParts.size > 1) {
+              throw new AnalysisException(
+                "Nested fields are not supported inside INSERT clauses of MERGE operations: " +
+                s"${nameParts.mkString("`", "`.`", "`")}")
+            }
+          }
+
+          val colNames = namePartsSeq.map(_.head)
+
+          // check there are no duplicates
+          val duplicateColNames = colNames.groupBy(identity).collect {
+            case (name, matchingNames) if matchingNames.size > 1 => name
+          }
+
+          if (duplicateColNames.nonEmpty) {
+            throw new AnalysisException(
+              "Duplicate column names inside INSERT clause: " +
+              duplicateColNames.mkString("[", ",", "]"))
+          }
+
+          // reorder assignments by the target table column order
+          i.copy(assignments = alignInsertActionAssignments(m.targetTable, colNames, assignments))
+        case _ =>
+          throw new AnalysisException("Not matched actions can only contain INSERT")
+      }
+
+      m.copy(matchedActions = alignedMatchedActions, notMatchedActions = alignedNotMatchedActions)
+  }
+
+  private def alignInsertActionAssignments(
+      targetTable: LogicalPlan,
+      insertCols: Seq[String],
+      assignments: Seq[Assignment]): Seq[Assignment] = {
+
+    val resolver = conf.resolver
+    targetTable.output.map { targetAttr =>
+      val assignment = assignments.find(a => resolver(targetAttr.name, getNameParts(a.key).head))
+      if (assignment.isEmpty) {
+        throw new AnalysisException(
+          s"Cannot find column '${targetAttr.name}' of the target table among " +
+          s"the INSERT columns: ${insertCols.mkString(", ")}. " +
+          "INSERT clauses must provide values for all columns of the target table.")
+      }
+
+      val key = assignment.get.key
+      val value = assignment.get.value
+      Assignment(key, castIfNeeded(value, key.dataType, resolver))
```

Comment:
Why not cast directly to the output column's data type?

Seems like this should cast to the types from `MergeIntoTable.output` so that all of the assignments produce the same type. Are all of the assignment keys for a column guaranteed to have the same output type?

Also, if column resolution adds a projection with a cast to the a table output type, can these expressions be optimized to remove the duplicate cast? If not, we may want to cast directly to the table's type and update the output of the `MergeIntoTable` node to be identical to the table so column resolution is done here rather than in the normal write resolution rule.

Reply:
I think you are correct we should cast to `targetAttr.dataType`. The resolution should be done before this rule so we should be safe with `castIfNeeded` not adding anything extra if it is not needed.

Reply:
Updated.

Reply:
Good context that this is a post-hoc rule, but I think that this might introduce problems.

Where does the output of `MergeIntoTable` come from? Table resolution is only going to add projections, so I don't think we can cast directly to the table type, we need to cast to the expected output type. I think that should be the same?

Reply:
I am not sure I fully understood. Could you check whether the current implementation is similar to what you mean?

Reply:
Will do.